### PR TITLE
Simulate infections

### LINKED ISSUE #509: forecast_infections
To bring `estimate_infections` into line with `forecast_secondary` and friends there should the the option of forecasting infections in `forecast_infections`.

This would work in the same way as `simualte_infections` but instead of taking a modified Rt time series it should simulate the generative process for the Rt model when not contained in the posterior predictions. This would decoupled forecasting from estimation. 

See `simulate_infections` here: https://epiforecasts.io/EpiNow2/reference/simulate_infections.html

### PR DESCRIPTION
**Pull request by sbfnk**
## Description

This PR closes #509 

It re-introduces the `simulate_infections` function, this time to simulate from given parameters (at the minimum, a trajectory of R and the number of initial infections). It also adds truncation to the simulation model (which had been missing).

Doing so required a small change to `extract_parameter_samples` which won't affect any existing functionality.

## Initial submission checklist

- [X] My PR is based on a package issue and I have explicitly linked it.
- [X] I have tested my changes locally (using `devtools::test()` and `devtools::check()`).
- [X] I have added or updated unit tests where necessary.
- [X] I have updated the documentation if required and rebuilt docs if yes (using `devtools::document()`).
- [X] I have followed the established coding standards (and checked using `lintr::lint_package()`).
- [X] I have added a news item linked to this PR.

## After the initial Pull Request 

- [x] I have reviewed Checks for this PR and addressed any issues as far as I am able.

### CODE REVIEW

File: R/simulate_infections.R [hunk 1]
```diff
@@ -1,19 +1,191 @@
...
-#' @return the result of [forecast_infections()]
-#' @export
-simulate_infections <- function(...) {
+##' Simulate infections using the renewal equation
+##'
+##' Simulations are done from given initial infections and, potentially
+##' time-varying, reproduction numbers. Delays and parameters of the observation
+##' model can be specified using the same options as in [estimate_infections()].
+##' @param R a data frame of reproduction numbers (column `R`) by date (column
+##'   `date`). R must be numric and date must be in date format. If any dates
```
Comment (jamesmbaazam) [inline R/simulate_infections.R:7]:
```suggestion
##'   `date`). R must be numeric and date must be in date format. If any dates
```

File: R/simulate_infections.R [hunk 2]
```diff
@@ -1,19 +1,191 @@
...
+        "This `estimates` option will be removed from [simulate_infections()] ",
+        "in version 2.1.0."
+      )
     )
-   forecast_infections(...)
+    return(forecast_infections(estimates = estimates, ...))
+  }
+
+  ## check inputs
+  assert_data_frame(R)
```
Comment (jamesmbaazam) [inline R/simulate_infections.R:64]: You can also check `any.missing` (default is TRUE), `min.rows`, `min.cols`, `ncols`, `col.names`.

File: R/simulate_infections.R [hunk 3]
```diff
@@ -1,19 +1,191 @@
...
+    must.include = c("date", "R")
+  )
+  assert_date(R$date, any.missing = FALSE)
+  assert_numeric(R$R, lower = 0)
+  assert_numeric(initial_infections, lower = 0)
+  assert_numeric(day_of_week_effect, lower = 0, null.ok = TRUE)
+  assert_numeric(pop, lower = 0)
+  assert_class(delays, "delay_opts")
+  assert_class(obs, "obs_opts")
+  assert_class(generation_time, "generation_time_opts")
```
Comment (jamesmbaazam) [inline R/simulate_infections.R:76]: We should probably bundle these checks into one as it's getting repeated in several places.

File: R/simulate_infections.R (hunk 2 above)
Comment (jamesmbaazam) [reply R/simulate_infections.R:64]: `any.missing` here will check for explicit missingness. I see implicit NA's are filled below.

File: R/simulate_infections.R [hunk 4]
```diff
@@ -1,19 +1,191 @@
...
+    delay = delays,
+    trunc = truncation
+  ))
+
+  if ((length(data$delay_mean_sd) > 0 && any(data$delay_mean_sd > 0)) ||
+      (length(data$delay_sd_sd) > 0 && any(data$delay_sd_sd > 0))) {
+    stop(
+      "Cannot simulate from uncertain parameters. Use the [fix_dist()] ",
+      "function to set the parameters of uncertain distributions either the ",
+      "mean or a randomly sampled value"
```
Comment (jamesmbaazam) [inline R/simulate_infections.R:116]: Should we make this explicit in the function description?

**Review by jamesmbaazam — COMMENTED** [review]
Thanks, Seb. Neat enhancement. I've left a few non-earth-shattering comments.

File: NEWS.md [hunk 5]
```diff
@@ -7,6 +7,7 @@
 * The `fixed` argument to `dist_spec` has been deprecated and replaced by a `fix_dist()` function. By @sbfnk in #503 and reviewed by @seabbs.
 * Updated `estimate_infections()` so that rather than imputing missing data, it now skips these data points in the likelihood. This is a breaking change as it alters the behaviour of the model when dates are missing from a time series but are known to be zero. We recommend that users check their results when updating to this version but expect this to in most cases improve performance. By @seabbs in #528 and reviewed by @sbfnk.
 * `simulate_infections` has been renamed to `forecast_infections` in line with `simulate_secondary` and `forecast_secondary`. The terminology is: a forecast is done from a fit to existing data, a simulation from first principles. By @sbfnk in #544 and reviewed by @seabbs.
+* A new `simulate_infections` function has been added that can be used to simulate from the model from given initial conditions and parameters. By @sbfnk in .
```
Comment (jamesmbaazam) [inline NEWS.md:10]:
```suggestion
* A new `simulate_infections` function has been added that can be used to simulate from the model from given initial conditions and parameters. By @sbfnk in #557.
```

File: R/simulate_infections.R (hunk 2 above)
Comment (sbfnk) [reply R/simulate_infections.R]: Thanks, improved in 57dd4e18

**Comment by sbfnk** [comment]
Thanks, @jamesmbaazam. Added your suggestions and a few tests which revealed few minor issues in the stan code which have now been fixed.

File: R/simulate_infections.R (hunk 3 above)
Comment (seabbs) [reply R/simulate_infections.R:84]: yes agree though that feels like its own issue?

File: R/simulate_infections.R [hunk 6]
```diff
@@ -1,19 +1,205 @@
...
+  ## remove any initial NAs
+  R <- R[!is.na(R)]
+
+  seeding_time <- get_seeding_time(delays, generation_time)
+  if (seeding_time > 1) {
+    ## estimate initial growth from initial reproduction number if seeding time
+    ## is greater than 1
+    initial_growth <- (R$R[1] - 1) / mean(generation_time)
+  } else {
+    initial_growth <- numeric(0)
```
Comment (seabbs) [inline R/simulate_infections.R:99]: can we add a test to cover this for completeness sake?

File: R/simulate_infections.R (hunk 3 above)
Comment (sbfnk) [reply R/simulate_infections.R:84]: It is now: #559

**Review by seabbs — DISMISSED** [review]
From an overall functionality perspective this looks good to me. One note about a testing gap that is somewhat optional. 

I think wait for @jamesmbaazam to be happy with the software dev side and then merge.

File: R/simulate_infections.R [hunk 7]
```diff
@@ -1,19 +1,205 @@
...
+  R <- merge.data.table(all_dates, R, by = "date", all.x = TRUE)
+  R <- R[, R := nafill(R, type = "locf")]
+  ## remove any initial NAs
+  R <- R[!is.na(R)]
+
+  seeding_time <- get_seeding_time(delays, generation_time)
+  if (seeding_time > 1) {
+    ## estimate initial growth from initial reproduction number if seeding time
+    ## is greater than 1
+    initial_growth <- (R$R[1] - 1) / mean(generation_time)
```
Comment (seabbs) [inline R/simulate_infections.R:97]: I don't think we are but we could be doing this in the inference model. In the Julia work Sam B implemented an improved version that doesn't use the mean approximation but I think this is likely good enough.

**Review by seabbs — DISMISSED** [review]

File: R/simulate_infections.R (hunk 7 above)
Comment (seabbs) [reply R/simulate_infections.R:97]: See: https://github.com/CDCgov/Rt-without-renewal/blob/aae49d72646546c43ebec5b638ee712ea1310a69/EpiAware/src/utilities.jl#L159

File: R/simulate_infections.R [hunk 8]
```diff
@@ -1,19 +1,205 @@
...
-#' Calling this function passes all arguments to [forecast_infections()]
-#' @description `r lifecycle::badge("deprecated")`
-#' @param ... Arguments to be passed to [forecast_infections()]
-#' @return the result of [forecast_infections()]
+#' Simulations are done from given initial infections and, potentially
+#' time-varying, reproduction numbers. Delays and parameters of the observation
+#' model can be specified using the same options as in [estimate_infections()].
+#'
+#' In order to simulate, all parameters that are specified such as the mean and
+#' standard deviatoin of delays or observation scaling, must be fixed.
```
Comment (jamesmbaazam) [inline R/simulate_infections.R:8]:
```suggestion
#' standard deviation of delays or observation scaling, must be fixed.
```

File: R/simulate_infections.R [hunk 9]
```diff
@@ -1,19 +1,205 @@
...
+#' Uncertain parameters are not allowed.
+#'
+#' A previous function called [simulate_infections()] that simulates from a
+#' given model fit has been renamed [forecast_infections()].
```
Comment (jamesmbaazam) [inline R/simulate_infections.R:12]: - Should we provide a quick justification here?
- Will this statement be removed after some time?

Comment (sbfnk) [reply R/simulate_infections.R]: Have added some context in cf0b772a

File: R/simulate_infections.R [hunk 10]
```diff
@@ -1,19 +1,205 @@
...
+  assert_numeric(R$R, lower = 0)
+  assert_numeric(initial_infections, lower = 0)
+  assert_numeric(day_of_week_effect, lower = 0, null.ok = TRUE)
+  assert_numeric(pop, lower = 0)
+  assert_class(delays, "delay_opts")
+  assert_class(obs, "obs_opts")
+  assert_class(generation_time, "generation_time_opts")
+
+  ## create R for all dates modelled
+  all_dates <- data.table(date = seq.Date(min(R$date), max(R$date), by = "day"))
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/simulate_infections.R:89]: Isn't it the other way round, i.e. `any.missing = FALSE` checks for no missing values?

``` r
R <- data.frame(x = c(NA_integer_, 2L))
checkmate::check_data_frame(R, any.missing = FALSE)
#> [1] "Contains missing values (column 'x', row 1)"
checkmate::check_data_frame(R, any.missing = TRUE)
#> [1] TRUE
```

<sup>Created on 2024-02-20 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Review by jamesmbaazam — DISMISSED** [review]
Thanks, Seb. LGTM.

File: R/simulate_infections.R (hunk 10 above)
[the remark this answers is not in the document]
Comment (jamesmbaazam) [reply R/simulate_infections.R:89]: Oh, yes. Sorry. We already check it in the R data.frame, so all good.

File: R/simulate_infections.R [hunk 11]
```diff
@@ -1,19 +1,205 @@
...
+#'
+#' In order to simulate, all parameters that are specified such as the mean and
+#' standard deviatoin of delays or observation scaling, must be fixed.
+#' Uncertain parameters are not allowed.
+#'
+#' A previous function called [simulate_infections()] that simulates from a
+#' given model fit has been renamed [forecast_infections()].
+#' @param R a data frame of reproduction numbers (column `R`) by date (column
+#'   `date`). R must be numeric and date must be in date format. If any dates
+#'   are missing, it will be assumed that R stays the same
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/simulate_infections.R]: Ah yes good point, will rephrase.

[the remark this answers is not in the document]
Comment (sbfnk) [reply R/simulate_infections.R]: a4713ec

[the remark this answers is not in the document]
Comment (jamesmbaazam) [reply R/simulate_infections.R]: Perfect.

**Pull request merged**